### PR TITLE
Add go-kit/log interface compatibility

### PR DESCRIPTION
**Pull request by fabxc**
@brian-brazil @grobie 

To some degree this is best effort I suppose. But practically this should work well for handing our log.Logger to any library using oklog or similar.

The lines a the middle are coming from tsdb.

screen shot 2017-06-08 at 17 12 04

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -149,13 +150,52 @@ type Logger interface {
...
+	Log(keyvals ...interface{}) error
+
 	With(key string, value interface{}) Logger
 }
 
 type logger struct {
 	entry *logrus.Entry
 }
 
+func (l logger) Log(keyvals ...interface{}) (err error) {
```
Comment (grobie) [inline log/log.go]: Remove named return value.

File: log/log.go [hunk 2]
```diff
@@ -149,13 +150,52 @@ type Logger interface {
...
 	Fatalf(string, ...interface{})
 
+	Log(keyvals ...interface{}) error
+
 	With(key string, value interface{}) Logger
 }
 
 type logger struct {
 	entry *logrus.Entry
 }
```
Comment (grobie) [inline log/log.go]: Missing doc string, should explain the behavior, the expectations of the input interface and what errors will be raised.

Comment (grobie) [inline log/log.go]: Should also mention that this implements go-kit/log

**Comment by brian-brazil** [comment]
The middle lines are lacking the source code line number, that's a pretty useful piece of information in an application log.

This PR also doesn't the change the fact that we'd still be using two different Go logging libraries. The number we should be using is exactly one.

**Comment by fabxc** [comment]
@brian-brazil there have been long discussions around how a standard Go logging interface should look like. @peterbourgon can tell you more.

The go-kit library requires implementing a single simple interface, which is trivial to do when having to plug an arbitrary logging library, like ours, into another package like prometheus/tsdb. If we were to use prometheus/common/log in prometheus/tsdb, this would be nearly impossible given the high amount of methods we require.

Showing the logged line can be helpful, but depending on your call nesting gives you results that aren't really useful. @peterbourgon, any way do work around that when using log/levels?
In general I think the logged message should be clear enough through usage of precise log key/vals and wrapping the error trace with pkg/errors.

**Comment by brian-brazil** [comment]
I'm okay with switching, after we've all discussed this and come to a consensus about switching. Until then we should be using the existing agreed library.

Last time around when we had the big discussion that resulted in common/log existing, having line numbers was seen as a critical feature. So it's not something to be dropped lightly.

I don't have skin in the game beyond having one system and having line numbers. Searching for where a log line came from without knowing the file and line number is quite irritating.

**Comment by peterbourgon** [comment]
@fabxc What is not working at the moment?

```go
package main

import (
	"os"

	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/log/level"
)

func main() {
	var logger log.Logger
	logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
	logger = level.NewFilter(logger, level.AllowWarn())
	logger = log.With(logger, "caller", log.DefaultCaller)
	logger = log.With(logger, "ts", log.DefaultTimestampUTC)

	foo(log.With(logger, "func", "foo"))
}

func foo(logger log.Logger) {
	level.Warn(logger).Log("hello", "world")
}
```

```
$ go run main.go
level=warn caller=main.go:21 ts=2017-06-19T09:54:12.208580733Z func=foo hello=world
```

**Comment by fabxc** [comment]
Weird! I had `log.DefaultCaller` resulting in `caller` being set to files of the log package. Maybe I wired it up incorrectly.

**Comment by peterbourgon** [comment]
The things that return log.Valuer typically need to go into a log.With statement, they aren't designed to stand alone in a logger.Log invocation. But glad that sorts you out :) The order of the keys in the log output is controlled by the order of the decorator declarations, by the way, so if you want e.g. timestamp-level-caller, it's also easy.

**Comment by fabxc** [comment]
Ah no, now I recall. The problem was that this log package attaches the source information right before printing. That's what I added here too and what then spit out oklog source files. That's more of a general problem of this package when it comes to plugging it into adapters.
I don't think logrus has the same capability as logrus to do that right thing here.

@brian-brazil  Can we merge this and I'll submit a PR to Prometheus integrating it. We'll wire up there to keep printing the source.

**Comment by brian-brazil** [comment]
> Can we merge this and I'll submit a PR to Prometheus integrating it. We'll wire up there to keep printing the source.

There's more to this than Prometheus. This is used in basically every other Go exporter too, and several 3rd party integrations.
We need consensus that this is the way we're going before we start doing all that work.

In any case we should probably finish the Cobra switch first, as that's a similarly large project that's just getting going.

**Comment by fabxc** [comment]
Nothing changes for any of those. So do you want to block this?
That means a) we switch all of TSDB to this log package, which I'll strongly oppose b) we do nothing and have TSDB just log for its own like it does now.

**Comment by brian-brazil** [comment]
I'm saying that there isn't presently a consensus on switching to go-kit for logging.
I'd suggest starting a thread on -dev with the aim of achieving consensus for your proposal to switch everything over.

If there is not consensus to switch, we would follow the standing consensus which would mean a).

My concern is primarily around having a homogeneous codebase, not which exact log library we use.

**Comment by brian-brazil** [comment]
This has all been done in other changes.

**Pull request closed without merging**